### PR TITLE
client/server decoupling: add protocol-definitions to driver-definitions

### PR DESCRIPTION
**Pull request by DLehenbauer**
This adds the types defined in 'protocol-definitions' to 'driver-definitions' and updates client dependencies/imports to use the 'driver-definitions' types.

This is a large review.  If using GitHub, I recommend:
a.  Use FireFox
b.  Use `document.querySelectorAll('.load-diff-button').forEach((button) => button.click());` to expand diffs

The approximate process used was:
1.  Copy types to src/protocol/*.ts
2.  S&R to update dependencies in package.json
3.  S&R to replace 'from "...protocol-definitions"' to 'from "...driver-definitions"'
4.  Use tools to clean up:
    a.  pnpm i
    b.  'flub modify fluid-imports --onlyInternal --packageRegex=driver-definitions'
    c.  npm run format
    d.  npm run policy-check:fix

Other notable changes:
* Required minor adjustments to layerInfo.json
* Semi-automatically expanded 'import * as api from ...' to regular imports for 'flub modify'.
* Needed to export 'ISequencedDocumentMessage' from 'fluid-framework' (not a new dependency, we just need to be explicit.)

### CODE REVIEW

File: azure/packages/azure-service-utils/src/generateToken.ts
Comment (DLehenbauer) [inline azure/packages/azure-service-utils/src/generateToken.ts:7]: Moving the types to 'driver-definitions' forces us to update imports to respect tagging.

This happened because 'protocol-definitions' has been tagged and trimmed, but we never published/consumed the latest version on the client, so the imports hadn't been updated to point to '/' vs. '/internal'.

Imports were updated using `flub modify fluid-imports --onlyInternal --packageRegex=driver-definitions`

File: azure/packages/azure-service-utils/src/test/types/validateAzureServiceUtilsPrevious.generated.ts
Comment (DLehenbauer) [inline azure/packages/azure-service-utils/src/test/types/validateAzureServiceUtilsPrevious.generated.ts:99]: This might be instability in 'fluid-type-test-generator'.  I'm waiting to see what CI says.

File: layerInfo.json
Comment (DLehenbauer) [inline layerInfo.json:22]: 'driver-definitions' no longer depends on 'protocol-definitions'

File: layerInfo.json
Comment (DLehenbauer) [inline layerInfo.json:69]: Telemetry-Utils now depends on driver-definitions instead of protocol-definitions.

File: layerInfo.json
Comment (DLehenbauer) [inline layerInfo.json:168]: Ditto for Azure-Service-Utils

File: packages/common/driver-definitions/src/index.ts
Comment (DLehenbauer) [inline packages/common/driver-definitions/src/index.ts:45]: Contents of 'protocol-definitions' are copied and re-exported verbatim.

File: packages/drivers/file-driver/api-report/file-driver.api.md
Comment (DLehenbauer) [inline packages/drivers/file-driver/api-report/file-driver.api.md:7]: The `flub modify fluid-imports` tool does not support the '* as api' pattern.

I semi-automatically expanded these into normal imports so that I could use `flub modify fluid-imports` to sort '/' vs '/internal' and so that we can use `flub modify` to adjust imports in case we retag anything in the future.

File: packages/utils/tool-utils/src/snapshotNormalizer.ts
Comment (alexvy86) [inline packages/utils/tool-utils/src/snapshotNormalizer.ts:11]: I see TreeEntry is an enum. No potential issues from using the same enum but defined in a different package, I imagine? (Like the same class from different packages would not pass `instanceof` checks).

File: pnpm-lock.yaml
Comment (alexvy86) [inline pnpm-lock.yaml:41428]: This feels funny here but no idea what it is.

File: experimental/PropertyDDS/packages/property-dds/package.json
Comment (alexvy86) [inline experimental/PropertyDDS/packages/property-dds/package.json:62]: driver-definitions can be removed from devDependencies now. Based on my review of the lockfile, I think that will be true for a few other packages.

File: packages/common/driver-definitions/src/protocol/clients.ts
Comment (alexvy86) [inline packages/common/driver-definitions/src/protocol/clients.ts:99]: I couldn't find any public exposure of ISequencedClient, just alpha ones... do we maybe need to review the tagging of these types now that we moved them here?

**Review by alexvy86 — APPROVED** [review]
Devtools changes are good. Had a look through some of the rest (particularly the "unique" bits like layering) and it looks reasonable to me.

File: packages/utils/tool-utils/src/snapshotNormalizer.ts
Comment (DLehenbauer) [reply packages/utils/tool-utils/src/snapshotNormalizer.ts:11]: Good thought.  TypeScript does not allow instanceof on enum types.  Comparisons between two enums is an error that would require a cast (no casts were added in this PR.)

...but the real answer is that we're working on ensuring the client types and server types never mix.  This PR gets as pretty close.

File: pnpm-lock.yaml
Comment (DLehenbauer) [reply pnpm-lock.yaml:41428]: Yeah...  It might be one of those packages that has a native node module for MacOS, but not other platforms.

I'm going to try reverting it by hand and see if CI is still happy.

File: experimental/PropertyDDS/packages/property-dds/package.json
Comment (DLehenbauer) [reply experimental/PropertyDDS/packages/property-dds/package.json:62]: Good catch.  There were four packages that had both normal and dev dependencies.  (Fixed).

File: packages/common/driver-definitions/src/protocol/clients.ts
Comment (DLehenbauer) [reply packages/common/driver-definitions/src/protocol/clients.ts:99]: Thanks for the heads up.  I'll have a look.

**Comment by Josmithr** [comment]
@DLehenbauer Is the goal here to remove `protocol-definitions` altogether? Or make it so server and client don't have to use the same package dependency for this stuff? I'm curious what the driving force for that is. IMO having a single package that defines type contracts between server and client seems reasonable. It should be small, and should probably _only_ export types, but that seems useful to me, even if it incurs some extra infra cost.

**Comment by tylerbutler** [comment]
Seems like the review would be smaller if we separately bumped protocol-defs in the client to a version with trimming. Then this PR would be only the driver-defs related changes.

**Review by Abe27342 — APPROVED** [review]
Changes to merge-tree, tree, sequence all look good. Didn't look deeply at the whole PR.

**Comment by DLehenbauer** [comment]
@tylerbutler - I'm sorry I didn't do the protocol-defs bump separately.  You're right, this could have been done in two steps.

Did you want to look more before I merge?

**Comment by DLehenbauer** [comment]
@Josmithr - Vlad and Tony are convinced that sharing types with server has historically had more downside than upside.

**Comment by tylerbutler** [comment]
> @tylerbutler - I'm sorry I didn't do the protocol-defs bump separately. You're right, this could have been done in two steps.
> 
> Did you want to look more before I merge?

No, if you have the review you need go ahead.

File: azure/packages/azure-service-utils/src/test/types/validateAzureServiceUtilsPrevious.generated.ts
Comment (tylerbutler) [reply azure/packages/azure-service-utils/src/test/types/validateAzureServiceUtilsPrevious.generated.ts:99]: This might be relevant: https://github.com/microsoft/FluidFramework/pull/20990#discussion_r1591318714 Also, not sure if you want to wait for that PR to merge before this one. I don't think it's necessary.

**Pull request merged**
